### PR TITLE
Salesforce update legacy actions

### LINKED ISSUE #2283: [UPDATE] Update Salesforce actions to ensure we're using >v30
It looks like [we're using v54.0](https://github.com/PipedreamHQ/pipedream/blob/master/components/salesforce_rest_api/salesforce_rest_api.app.mjs#L78) in the latest triggers but v20.0 in some of the actions (and test request), which [Salesforce is deprecating soon](https://help.salesforce.com/s/articleView?id=000354473&type=1).

Let's update so everything is on the latest version and consistent.

### PR DESCRIPTION
**Pull request by andrewjschuang**
-- **UPDATE** --

Made all the refactoring changes here in this PR. This should be reviewed as one PR for the related issue.

- Updated actions and sources API to v50.0
- Created props for each Salesforce SObject in separated files in `./common/sobjects/`
- Required props for each SObject are inside the component and optional props are in `./common/sobjects/`
- `fieldSelector: string[]` prop for selecting which optional props are going to be used, each one loaded as an additional prop
- moved async options for `objectType` prop
- bumped minor version for all actions and sources

---

This PR addressed issue #2283.

Salesforce legacy actions were using API v20.0 that is deprecated and will soon be retired. This PR updates the API version to v50.0, staying consistent with the sources. Also some minor refactoring was made, though not enough to be compliant with the guidelines.

~~Therefore, this PR should not be merged alone.~~

@dannyroosevelt Should a new card be created for refactoring these actions? I think combining them into one PR would be too big of a change, especially for the review. Also, would there be a big impact if most of the props were to be deleted? There is a large number of props in each action and I think refactoring each one would take a very long time.

Resolves #2283

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

**Comment by dannyroosevelt** [comment]
> This PR addressed issue #2283.
> 
> Salesforce legacy actions were using API v20.0 that is deprecated and will soon be retired. This PR updates the API version to v50.0, staying consistent with the sources. Also some minor refactoring was made, though not enough to be compliant with the guidelines.
> 
> **Therefore, this PR should not be merged alone.**
> 
> @dannyroosevelt Should a new card be created for refactoring these actions? I think combining them into one PR would be too big of a change, especially for the review. Also, would there be a big impact if most of the props were to be deleted? There is a large number of props in each action and I think refactoring each one would take a very long time.
> 
> Resolves #2283

@andrewjschuang thank you for the detailed and clear update! If you think it makes sense to break out the work into multiple PRs, that makes sense and sounds good to me. I think probably just keep a single issue to track all the work on the GitHub project board, but you can attach multiple PRs.

And for the props question, that's fine to get rid of a lot of them (there are too many there anyway, and the current component actions were converted programmatically from the legacy actions we had in the v1 builder). I think it makes sense to **rev the `major` version for any component you're updating**, since there are major changes. And that will notify users in the UI about breaking changes when they do an update.

How does that all sound?

**Comment by andrewjschuang** [comment]
> > This PR addressed issue #2283.
> > Salesforce legacy actions were using API v20.0 that is deprecated and will soon be retired. This PR updates the API version to v50.0, staying consistent with the sources. Also some minor refactoring was made, though not enough to be compliant with the guidelines.
> > **Therefore, this PR should not be merged alone.**
> > @dannyroosevelt Should a new card be created for refactoring these actions? I think combining them into one PR would be too big of a change, especially for the review. Also, would there be a big impact if most of the props were to be deleted? There is a large number of props in each action and I think refactoring each one would take a very long time.
> > Resolves #2283
> 
> @andrewjschuang thank you for the detailed and clear update! If you think it makes sense to break out the work into multiple PRs, that makes sense and sounds good to me. I think probably just keep a single issue to track all the work on the GitHub project board, but you can attach multiple PRs.
> 
> And for the props question, that's fine to get rid of a lot of them (there are too many there anyway, and the current component actions were converted programmatically from the legacy actions we had in the v1 builder). I think it makes sense to **rev the `major` version for any component you're updating**, since there are major changes. And that will notify users in the UI about breaking changes when they do an update.
> 
> How does that all sound?

Alright, sounds great! Thanks

File: components/salesforce_rest_api/actions/salesforce-update-contact/salesforce-update-contact.mjs [hunk 1]
```diff
@@ -1,345 +1,56 @@
...
-      type: "string",
-      description: "References the company's ID in Data.com. If an account has a value in this field, it means that the account was imported from Data.com. If the field value is null, the account was not imported from Data.com. Maximum size is 20 characters. Available in API version 22.0 and later. Label is Data.com Key. Important The Jigsaw field is exposed in the API to support troubleshooting for import errors and reimporting of corrected data. Do not modify this value.",
-      optional: true,
-    },
     LastName: {
       type: "string",
+      label: "LastName",
       description: "Required. Last name of the contact up to 80 characters.",
       optional: true,
     },
```
Comment (michelle0927) [inline components/salesforce_rest_api/actions/salesforce-update-contact/salesforce-update-contact.mjs:28]: Is `LastName` required or optional? The description says required, but it's listed as optional.

File: components/salesforce_rest_api/salesforce_rest_api.app.mjs [hunk 2]
```diff
@@ -6,26 +6,37 @@ export default {
...
-          eventType,
-        } = context;
+        const { page } = context;
         if (page !== 0) {
-          // The list of allowed SObject types is static and exhaustively
-          // provided through a single method call
-          return [];
+          return {
+            options: [],
+          };
```
Comment (michelle0927) [inline components/salesforce_rest_api/salesforce_rest_api.app.mjs:31]: We can just return the empty array.
```suggestion
          return [];
```

File: components/salesforce_rest_api/salesforce_rest_api.app.mjs [hunk 3]
```diff
@@ -6,26 +6,37 @@ export default {
...
+        const { sobjects } = await this.listSObjectTypes();
+        const options = sobjects
+          .filter((sobject) => sobject.replicateable)
+          .map((sobject) => ({
+            label: sobject.label,
+            value: sobject.name,
+          }));
         return {
           options,
         };
```
Comment (michelle0927) [inline components/salesforce_rest_api/salesforce_rest_api.app.mjs:42]: We don't need to return `options` inside an object.
```suggestion
        const { sobjects } = await this.listSObjectTypes();
        return sobjects
          .filter((sobject) => sobject.replicateable)
          .map((sobject) => ({
            label: sobject.label,
            value: sobject.name,
          }));
```

File: components/salesforce_rest_api/sources/common-instant.mjs [hunk 4]
```diff
@@ -12,17 +12,26 @@ export default {
...
+        }
+
+        const supportedObjectTypes = this.salesforce.listAllowedSObjectTypes(this.getEventType());
+        const options = supportedObjectTypes.map((i) => ({
+          label: i.label,
+          value: i.name,
+        }));
+        return {
+          options,
+        };
```
Comment (michelle0927) [inline components/salesforce_rest_api/sources/common-instant.mjs:34]: We don't need to return the options inside an object.
```suggestion
        return supportedObjectTypes.map((i) => ({
          label: i.label,
          value: i.name,
        }));
```

**Review by michelle0927 — CHANGES_REQUESTED** [review]
I don't think the action filenames need to start with "salesforce". I think it would be better to rename them as `create-account.mjs`, `create-attachment.mjs`, etc. by renaming the folder names, filenames, and component `key`s.

**Review by michelle0927 — APPROVED** [review]

**Comment by DilanAthukorala** [comment]
Hi @andrewjschuang, I'm getting the following error when trying add Salesforce Account to Pipedream

Pipedream failed to connect to your account.

Please reach out to integration-errors@pipedream.com and provide this error ID: 18cb5e9b3c294adf8700706996654b33

https://api.pipedream.com/connect/oauth/oa_GXxiEz/callback?code=aPrx_yTkbIprMMiwU2cLnIHnyqFTdWteO_jvKP0lhphtuXyT7f3JhO11tiiqKLnC.TziasSmzQ%3D%3D&state=dV9xbGhwRURMfGFjNjg0OTJlNDQ1OTc4MzNkODI2MGFlNWFmMDliZTFj

**Comment by dannyroosevelt** [comment]
Here's the error we got for you @DilanAthukorala -- are you able to enable the Rest API for your account?
```
[
  {
    "message": "The REST API is not enabled for this Organization.",
    "errorCode": "API_DISABLED_FOR_ORG"
  }
]
```

**Comment by vunguyenhung** [comment]
Hello everyone, I have tested this PR and there're some test cases failed or needed improvement. 

Please check the test report below for more information
[Salesforce_Test_Report_2557.pdf](https://github.com/PipedreamHQ/pipedream/files/8687318/Salesforce_Test_Report_2557.pdf)

**Comment by dannyroosevelt** [comment]
What's the status of this?

**Comment by andrewjschuang** [comment]
> What's the status of this?

Pending #1474 because of conflicts.

**Pull request closed without merging**